sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,232 @@
+  @VisibleForTesting ScheduledFuture<?> flushFuture;
+  private long byteCounter;
+
+  public BeamFnDataOutboundAggregator(
+      PipelineOptions options, StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.sizeLimit = getSizeLimit(options);
+    this.timeLimit = getTimeLimit(options);
+    this.outputLocations = new HashMap<>();
+    this.buffer = new ConcurrentHashMap<>();
+    this.outboundObserver = outboundObserver;
+    if (timeLimit > 0) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    } else {
+      this.flushFuture = null;
+    }
+  }
+
+  /**
+   * Register the outbound logical endpoint alongside it's value coder. All registered endpoints
+   * will eventually be removed when {@link #close()} is called at the end of a bundle.
+   */
+  public <T> void registerOutputLocation(LogicalEndpoint endpoint, Coder<T> coder) {
+    LOG.debug("Registering endpoint: {}", endpoint);
+    outputLocations.put(endpoint, coder);
+  }
+
+  public void flush() throws IOException {
+    if (byteCounter > 0) {
+      outboundObserver.onNext(convertBufferForTransmission().build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  @Override
+  public void close() {
+    LOG.debug("Closing streams for outbound endpoints {}", outputLocations);
+    if (byteCounter == 0 && outputLocations.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    for (LogicalEndpoint outputLocation : outputLocations.keySet()) {
+      if (outputLocation.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setTimerFamilyId(outputLocation.getTimerFamilyId())
+            .setIsLast(true);
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setIsLast(true);
+      }
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputLocations.clear();
+  }
+
+  public <T> void accept(LogicalEndpoint endpoint, T data) throws Exception {
+    if (timeLimit > 0) {
+      checkFlushThreadException();
+    }
+    Output output = buffer.computeIfAbsent(endpoint, e -> ByteString.newOutput());
+    int size = output.size();
+    ((Coder<T>) outputLocations.get(endpoint)).encode(data, output);
+    byteCounter += output.size() - size;
```

Comment:
All elements must encode to be at least one byte in length to ensure that we can't enter an infinite loop when encoding/decoding since we can encode/decode an infinite number of 0 byte length objects.

See https://github.com/apache/beam/blob/e38c6ac5760f55494e71f61d0271a624ca96ea17/sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DataStreams.java#L239

Was this logic missing in the existing implementation?

Reply:
done.